File: src/main/java/org/capturecoop/cccolorutils/chooser/gui/CCColorChooserSinglePreview.java
package org.capturecoop.cccolorutils.chooser.gui;

import org.capturecoop.cccolorutils.CCColor;
import org.capturecoop.cccolorutils.chooser.CCColorChooser;

import javax.swing.*;
import java.awt.*;
import java.awt.image.BufferedImage;

public class CCColorChooserSinglePreview extends JPanel {
    private final CCColor color;
    private final BufferedImage previewBackground;

    public CCColorChooserSinglePreview(CCColorChooser colorChooser, BufferedImage previewBackground) {
        color = colorChooser.getColor();
        color.addChangeListener(e -> repaint());
        this.previewBackground = previewBackground;
    }

    @Override
    public void paint(Graphics g) {
        super.paint(g);
        if(previewBackground != null)
            g.drawImage(previewBackground.getSubimage(0, 0, getHeight(), getHeight()), getWidth() / 2 - getHeight()/2, 0, getHeight(), getHeight(), null);
        if(color != null) {
            g.setColor(color.getPrimaryColor());
            g.fillRect(getWidth() / 2 - getHeight()/2, 0, getHeight(), getHeight());
        }
    }
}

File: src/main/kotlin/org/capturecoop/cccolorutils/chooser/gui/CCColorChooserSinglePreview.kt
package org.capturecoop.cccolorutils.chooser.gui

import org.capturecoop.cccolorutils.CCColor
import org.capturecoop.cccolorutils.chooser.CCColorChooser
import java.awt.Graphics
import java.awt.image.BufferedImage
import javax.swing.JPanel
import javax.swing.event.ChangeListener

class CCColorChooserSinglePreview(colorChooser: CCColorChooser, private val previewBackground: BufferedImage?) : JPanel() {
    private val color: CCColor = colorChooser.color

    init {
        color.changeListeners.add(ChangeListener { repaint() })
    }

    override fun paint(g: Graphics) {
        super.paint(g)
        val p = width / 2 - height / 2
        previewBackground?.let { g.drawImage(it.getSubimage(0, 0, height, height), p, 0, height, height, null) }
        g.color = color.primaryColor
        g.fillRect(p, 0, height, height)
    }
}